File: 0315 - Count of Smaller Numbers After Self/solution.py
class BIT:
    def __init__(self, sz):
        self.sz = sz
        self.v = [0]*(sz+1)
    
    def update(self, x, d):
        x += 1
        while x <= self.sz:
            self.v[x] += d
            x += x&-x
            
    def query(self, x):
        x += 1; res = 0
        while x > 0:
            res += self.v[x]
            x -= x&-x
        return res
    
class Solution:
    def countSmaller(self, A: List[int]) -> List[int]:
        n, BASE, b = len(A), 10000, BIT(20001)
        for i in range(n-1, -1, -1):
            val = A[i]
            A[i] = b.query(BASE+(val-1))
            b.update(BASE+val, 1)
        return A

File: 0315 - Count of Smaller Numbers After Self/solution.cpp
#define P pair<int, int>
#define F first
#define S second
class Solution {
private:
    void mergeSort(vector<P>& p, vector<int>& res){
        int n = p.size(), mid = n/2;
        if(n < 2) return;
        vector<P> left(p.begin(), p.begin()+mid), right(p.begin()+mid, p.end());
        mergeSort(left, res); 
        mergeSort(right, res);
        for(int i = n-1; i >= 0; i--){
            if(right.empty() || !left.empty()&&left.back().F>right.back().F)
                p[i] = left.back(), left.pop_back(), res[p[i].S] += right.size();
            else
                p[i] = right.back(), right.pop_back();
        }
    }
public:
    vector<int> countSmaller(vector<int>& A) {
        int n = A.size();
        vector<P> p(n);
        for(int i = 0; i < n; i++) 
            p[i] = {A[i], i}, A[i] = 0;
        mergeSort(p, A);
        return A;
    }
};